File: app/services/error_tracking_service.rb
# frozen_string_literal: true

# ErrorTrackingService - Unified error tracking with Sentry integration
#
# Provides structured error logging with consistent categorization,
# stack traces, and optional Sentry reporting.
#
# Usage:
#   ErrorTrackingService.capture(exception, context: { contact_id: 123 })
#   ErrorTrackingService.warn("Rate limit exceeded", tags: { api: :twilio })
#
class ErrorTrackingService
  # Error categories for consistent classification
  CATEGORIES = {
    transient: %w[timeout network rate_limit 503 502 429],
    permanent: %w[invalid_number not_found authentication validation],
    configuration: %w[missing_key no_credentials disabled],
    internal: %w[nil_class undefined_method]
  }.freeze

  class << self
    # Capture an exception with full context
    #
    # @param exception [Exception] The exception to capture
    # @param context [Hash] Additional context (contact_id, job_id, etc.)
    # @param tags [Hash] Tags for categorization
    # @param level [Symbol] :error, :warning, :info
    def capture(exception, context: {}, tags: {}, level: :error)
      category = categorize_exception(exception)

      log_structured(
        level: level,
        category: category,
        exception_class: exception.class.name,
        message: exception.message,
        context: context,
        tags: tags,
        backtrace: exception.backtrace&.first(10)&.join("\n")
      )

      # Report to Sentry if available and not transient
      return unless defined?(Sentry) && category != :transient

      Sentry.capture_exception(exception, extra: context, tags: tags.merge(category: category))
    end

    # Log a warning with structured format
    def warn(message, context: {}, tags: {})
      log_structured(
        level: :warn,
        category: :warning,
        message: message,
        context: context,
        tags: tags
      )
    end

    # Log info with structured format
    def info(message, context: {}, tags: {})
      log_structured(
        level: :info,
        category: :info,
        message: message,
        context: context,
        tags: tags
      )
    end

    # Track a rate limit event
    def track_rate_limit(provider:, retry_after: nil, context: {})
      log_structured(
        level: :warn,
        category: :rate_limit,
        message: "Rate limit exceeded for #{provider}",
        context: context.merge(retry_after: retry_after),
        tags: { provider: provider, rate_limited: true }
      )

      # Optionally report to Sentry as warning
      return unless defined?(Sentry)

      Sentry.capture_message(
        "Rate limit: #{provider}",
        level: :warning,
        extra: context.merge(retry_after: retry_after)
      )
    end

    # Track a circuit breaker event
    def track_circuit_breaker(service:, state:, context: {})
      level = state == :open ? :error : :info

      log_structured(
        level: level,
        category: :circuit_breaker,
        message: "Circuit breaker #{state} for #{service}",
        context: context,
        tags: { service: service, circuit_state: state }
      )

      return unless state == :open && defined?(Sentry)

      Sentry.capture_message(
        "Circuit breaker OPEN: #{service}",
        level: :error,
        extra: context
      )
    end

    private

    def categorize_exception(exception)
      message = exception.message.to_s.downcase

      CATEGORIES.each do |category, keywords|
        return category if keywords.any? { |kw| message.include?(kw) }
      end

      # Default categorization by exception type
      case exception
      when Timeout::Error, Net::OpenTimeout, Net::ReadTimeout
        :transient
      when ArgumentError, TypeError
        :permanent
      when ActiveRecord::RecordInvalid, ActiveRecord::RecordNotFound
        :permanent
      else
        :unknown
      end
    end

    def log_structured(level:, category:, message:, context: {}, tags: {}, backtrace: nil)
      log_entry = {
        timestamp: Time.current.iso8601,
        level: level.to_s.upcase,
        category: category,
        message: message,
        context: context,
        tags: tags
      }

      log_entry[:backtrace] = backtrace if backtrace.present?

      # Format for structured logging (JSON in production, readable in dev)
      if Rails.env.production?
        Rails.logger.send(level, log_entry.to_json)
      else
        formatted = "[#{level.to_s.upcase}] [#{category}] #{message}"
        formatted += " | context: #{context.inspect}" if context.present?
        formatted += " | tags: #{tags.inspect}" if tags.present?
        formatted += "\n  #{backtrace}" if backtrace.present?
        Rails.logger.send(level, formatted)
      end
    end
  end
end

File: laravel-app/app/Services/ErrorTrackingService.php
<?php

namespace App\Services;

use Illuminate\Support\Facades\Log;
use Exception;

/**
 * ErrorTrackingService - Unified error tracking with Sentry integration
 *
 * Provides structured error logging with consistent categorization,
 * stack traces, and optional Sentry reporting.
 *
 * Usage:
 *   ErrorTrackingService::capture($exception, ['contact_id' => 123]);
 *   ErrorTrackingService::warn("Rate limit exceeded", ['api' => 'twilio']);
 */
class ErrorTrackingService
{
    /**
     * Error categories for consistent classification
     */
    const CATEGORIES = [
        'transient' => ['timeout', 'network', 'rate_limit', '503', '502', '429'],
        'permanent' => ['invalid_number', 'not_found', 'authentication', 'validation'],
        'configuration' => ['missing_key', 'no_credentials', 'disabled'],
        'internal' => ['nil_class', 'undefined_method'],
    ];

    /**
     * Capture an exception with full context
     *
     * @param Exception $exception The exception to capture
     * @param array $context Additional context (contact_id, job_id, etc.)
     * @param array $tags Tags for categorization
     * @param string $level 'error', 'warning', 'info'
     * @return void
     */
    public static function capture(Exception $exception, array $context = [], array $tags = [], string $level = 'error'): void
    {
        $category = self::categorizeException($exception);

        self::logStructured([
            'level' => $level,
            'category' => $category,
            'exception_class' => get_class($exception),
            'message' => $exception->getMessage(),
            'context' => $context,
            'tags' => $tags,
            'backtrace' => implode("\n", array_slice($exception->getTrace(), 0, 10)),
        ]);

        // Report to Sentry if available and not transient
        if ($category !== 'transient' && function_exists('app') && app()->bound('sentry')) {
            if (function_exists('\Sentry\captureException')) {
                \Sentry\captureException($exception);
            }
        }
    }

    /**
     * Log a warning with structured format
     *
     * @param string $message
     * @param array $context
     * @param array $tags
     * @return void
     */
    public static function warn(string $message, array $context = [], array $tags = []): void
    {
        self::logStructured([
            'level' => 'warn',
            'category' => 'warning',
            'message' => $message,
            'context' => $context,
            'tags' => $tags,
        ]);
    }

    /**
     * Log info with structured format
     *
     * @param string $message
     * @param array $context
     * @param array $tags
     * @return void
     */
    public static function info(string $message, array $context = [], array $tags = []): void
    {
        self::logStructured([
            'level' => 'info',
            'category' => 'info',
            'message' => $message,
            'context' => $context,
            'tags' => $tags,
        ]);
    }

    /**
     * Track a rate limit event
     *
     * @param string $provider
     * @param int|null $retryAfter
     * @param array $context
     * @return void
     */
    public static function trackRateLimit(string $provider, ?int $retryAfter = null, array $context = []): void
    {
        $contextWithRetry = array_merge($context, ['retry_after' => $retryAfter]);

        self::logStructured([
            'level' => 'warn',
            'category' => 'rate_limit',
            'message' => "Rate limit exceeded for {$provider}",
            'context' => $contextWithRetry,
            'tags' => ['provider' => $provider, 'rate_limited' => true],
        ]);

        // Optionally report to Sentry as warning
        if (function_exists('\Sentry\captureMessage')) {
            \Sentry\captureMessage(
                "Rate limit: {$provider}",
                \Sentry\Severity::warning()
            );
        }
    }

    /**
     * Track a circuit breaker event
     *
     * @param string $service
     * @param string $state
     * @param array $context
     * @return void
     */
    public static function trackCircuitBreaker(string $service, string $state, array $context = []): void
    {
        $level = $state === 'open' ? 'error' : 'info';

        self::logStructured([
            'level' => $level,
            'category' => 'circuit_breaker',
            'message' => "Circuit breaker {$state} for {$service}",
            'context' => $context,
            'tags' => ['service' => $service, 'circuit_state' => $state],
        ]);

        if ($state === 'open' && function_exists('\Sentry\captureMessage')) {
            \Sentry\captureMessage(
                "Circuit breaker OPEN: {$service}",
                \Sentry\Severity::error()
            );
        }
    }

    /**
     * Categorize exception
     */
    private static function categorizeException(Exception $exception): string
    {
        $message = strtolower($exception->getMessage());

        foreach (self::CATEGORIES as $category => $keywords) {
            foreach ($keywords as $keyword) {
                if (strpos($message, $keyword) !== false) {
                    return $category;
                }
            }
        }

        // Default categorization by exception type
        $exceptionClass = get_class($exception);

        if (strpos($exceptionClass, 'Timeout') !== false || strpos($exceptionClass, 'Network') !== false) {
            return 'transient';
        }

        if (strpos($exceptionClass, 'Invalid') !== false || strpos($exceptionClass, 'NotFound') !== false) {
            return 'permanent';
        }

        return 'unknown';
    }

    /**
     * Log structured data
     */
    private static function logStructured(array $data): void
    {
        $logEntry = [
            'timestamp' => now()->toIso8601String(),
            'level' => strtoupper($data['level'] ?? 'INFO'),
            'category' => $data['category'] ?? 'unknown',
            'message' => $data['message'] ?? '',
            'context' => $data['context'] ?? [],
            'tags' => $data['tags'] ?? [],
        ];

        if (!empty($data['backtrace'])) {
            $logEntry['backtrace'] = $data['backtrace'];
        }

        if (!empty($data['exception_class'])) {
            $logEntry['exception_class'] = $data['exception_class'];
        }

        // Format for structured logging (JSON in production, readable in dev)
        if (app()->environment('production')) {
            Log::log(
                strtolower($data['level'] ?? 'info'),
                json_encode($logEntry)
            );
        } else {
            $formatted = "[{$logEntry['level']}] [{$logEntry['category']}] {$logEntry['message']}";

            if (!empty($logEntry['context'])) {
                $formatted .= " | context: " . json_encode($logEntry['context']);
            }

            if (!empty($logEntry['tags'])) {
                $formatted .= " | tags: " . json_encode($logEntry['tags']);
            }

            if (!empty($logEntry['backtrace'])) {
                $formatted .= "\n  {$logEntry['backtrace']}";
            }

            Log::log(strtolower($data['level'] ?? 'info'), $formatted);
        }
    }
}
